doc/adding_features.md

```diff
@@ -0,0 +1,161 @@
+
+To render our new `RandomString` component, we can simply import it and add an instantiating JSX tag to the render method of another component in our app: `<RandomString />`. If our component tooks props (like arguments), this is where we'd pass them in.
+
+Then we add two lines to our main reducer file to import our new reducer from `randomString.js` and add it to our root reducer (`static/js/reducers.js`):
+
+``` javascript
+...
+import randomStringReducer from './ducks/randomString';
+
+
+const root = combineReducers({
+  ...,
+  randomString: randomStringReducer
+});
+...
+```
+
+
+## Back-end
+
+Now let's take a look at what we'll need to add to the back-end to make our new feature work.
+
+We'll start by defining a new request handler in a new file located in `skyportal/handlers/api/random_string.py`. This handler will be a subclass of `BaseHandler`, a handler based on Tornado's HTTP request handler defined in `skyportal/handlers/base.py`.
+
+The decorator `@auth_or_token` tells the application that the request must come from either or a logged in user (via the browser) or must include a valid token in the request header. If neither of these are true, the request returns with an error.
+
+`skyportal/handlers/api/random_string.py`:
+``` python
+import uuid
+from baselayer.app.access import auth_or_token
+from ..base import BaseHandler
+
+
+class RandomStringHandler(BaseHandler):
+    @auth_or_token
+    def get(self):
+        return self.success(data={'value': str(uuid.uuid4())})
+```
+After adding the relevant import lines to `skyportal/handlers/api/__init__.py` and `skyportal/app_server.py` to import our new `RandomStringHandler`, we then add an element to our endpoint & handlers list in `skyportal/app_server.py` to map the associated URL endpoint to the new handler, which is a tuple consisting of a regex string describing the matching URL pattern, and the handler object: `(r'/api/random_string', RandomStringHandler)`. See the `app_server.py` file for other examples.
+
+Now our new feature should be fully implemented, and we'll add a test to make sure it's working as expected. We use Selenium in our front-end tests; we'll use that to make sure a UUID string appears after (and not before) we click the button. This test assumes we've rendered our new component inside the `Profile` component, which is rendered when a user navigates to `/profile` (this is arbitrary - adjust these to reflect whichever location you choose to render your component).
+
+`skyportal/tests/frontend/test_random_string.py`:
+
+``` python
+import pytest
+import re
+from selenium.webdriver.support.ui import Select
+
+def test_random_string_generator_button(driver, user):
+    driver.get(f'/become_user/{user.id}')
```

Comment:
You explained this to me today, but I may need another round of explanation.  Can we not make the token belong to the same user, so that it is unnecessary to become the user?  Otherwise, we could probably automatically log the correct user in anyway.